### PR TITLE
crash tester test suite

### PR DESCRIPTION
**Pull request by weissi**
Motivation:

It's often very important to test that in certain cases we actually
crash with a certain error message whilst having access to all of NIO.

Modifications:

Add a crash tester test suite.

Result:

We can test crashers.

### CODE REVIEW

File: Sources/NIOCrashTester/CrashTests.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import Foundation
+
+let group = MultiThreadedEventLoopGroup(numberOfThreads: 2)
+
+struct Foo {
+    let multiThreadedELGCrashesOnZeroThreads = CrashTest { requireCrashRegex in
```
Comment (weissi) [inline Sources/NIOCrashTester/CrashTests.swift:21]: note to reviewers: this is how crash tests look. For each thing you'd like to test you add a

```swift
let whatYouWantToTest = CrashTest { requireCrashRegex in
    requireCrashRegex("^regex of what message you want to see")

    whatMakesItCrash()
}
```

File: Sources/NIOCrashTester/EventLoopCrashTests.swift [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+fileprivate let group = MultiThreadedEventLoopGroup(numberOfThreads: 2)
+
+struct EventLoopCrashTests {
+    let testMultiThreadedELGCrashesOnZeroThreads = CrashTest(
+        regex: "^Precondition failed: numberOfThreads must be positive"
+    ) {
+        try? MultiThreadedEventLoopGroup(numberOfThreads: 0).syncShutdownGracefully()
+    }
+
+    let testWaitCrashesWhenOnEL = CrashTest(
```
Comment (weissi) [inline Sources/NIOCrashTester/EventLoopCrashTests.swift:27]: note to reviewers: these files are the core. To add a new crash test, the only thing you need to do is

```
let testWhatYouAreTesting = CrashTest(regex: "^outputYouExpect") {
    codeThatWillCrash()
}
```

Comment (weissi) [reply Sources/NIOCrashTester/EventLoopCrashTests.swift:27]: the will automatically be collected through the `Mirror` of the test struct. If you add a new test suite, then you'll need to edit `main.swift`.

File: Sources/NIOCrashTester/CrashTests+EventLoop.swift [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    let testPreconditionInEventLoop = CrashTest(
+        regex: "^Precondition failed: file .*CrashTests\\+EventLoop.swift, line \(#line + 2)"
+    ) {
+        group.next().preconditionInEventLoop()
+    }
+
+    let testAssertNotInEventLoop = CrashTest(
+        regex: "^Precondition failed: file .*CrashTests\\+EventLoop.swift, line \(#line + 4)"
```
Comment (weissi) [inline Sources/NIOCrashTester/CrashTests+EventLoop.swift:49]: CC @gwynne / @Davidde94

File: Sources/NIOCrashTester/CrashTests+ByteBuffer.swift [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import Foundation
```
Comment (glbrntt) [inline Sources/NIOCrashTester/CrashTests+ByteBuffer.swift:16]: Seems unnecessary? (and elsewhere)

File: Sources/NIOCrashTester/CrashTests+EventLoop.swift [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        regex: #"^Precondition failed: BUG DETECTED: wait\(\) must not be called when on an EventLoop"#
+    ) {
+        let promise = group.next().makePromise(of: Void.self)
+        try? group.next().submit {
+            try? promise.futureResult.wait()
+        }.wait()
+    }
+
+    let testAssertInEventLoop = CrashTest(
+        regex: "^Precondition failed: file .*CrashTests\\+EventLoop.swift, line \(#line + 2)"
```
Comment (glbrntt) [inline Sources/NIOCrashTester/CrashTests+EventLoop.swift:37]: Super minor, this feels slightly brittle setting the expected line like this, would it be better to just manually override the line in `assertInEventLoop` etc.?

File: Sources/NIOCrashTester/CrashTests+ByteBuffer.swift [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+fileprivate let group = MultiThreadedEventLoopGroup(numberOfThreads: 2)
```
Comment (Lukasa) [inline Sources/NIOCrashTester/CrashTests+ByteBuffer.swift:18]: We don't seem to be using this at all.

File: Sources/NIOCrashTester/CrashTests+ByteBuffer.swift (hunk 4 above)
Comment (ktoso) [reply Sources/NIOCrashTester/CrashTests+ByteBuffer.swift:16]: (There isn't a linter which could fail if an import isn't used, is there? If there is I'd love to know -- Foundation imports creeping into more files than necessary is a bit annoying to me in my work :-/)

File: Sources/NIOCrashTester/CrashTests+EventLoop.swift (hunk 5 above)
Comment (weissi) [reply Sources/NIOCrashTester/CrashTests+EventLoop.swift:36]: good idea!

Comment (weissi) [reply Sources/NIOCrashTester/CrashTests+EventLoop.swift]: done

File: Sources/NIOCrashTester/CrashTests+ByteBuffer.swift (hunk 6 above)
Comment (weissi) [reply Sources/NIOCrashTester/CrashTests+ByteBuffer.swift]: thanks, removed

File: Sources/NIOCrashTester/CrashTests+ByteBuffer.swift (hunk 4 above)
Comment (weissi) [reply Sources/NIOCrashTester/CrashTests+ByteBuffer.swift]: removed the unused imports

Comment (ktoso) [reply Sources/NIOCrashTester/CrashTests+ByteBuffer.swift]: I see so you test it based on what's linked hm... Was hoping there's some linter that per source could check it and I could "permit" it sometimes. No worries, just asking

**Comment by Davidde94** [comment]
This is really cool :)

**Pull request merged**